libp2p/services/autonatservice.nim

```diff
@@ -0,0 +1,128 @@
+# those terms.
+
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import std/[options, tables]
+import chronos
+import ../switch
+import ../protocols/[connectivity/autonat,
+                    rendezvous]
+import ../protocols/connectivity/relay/[relay, client]
+import ../discovery/[rendezvousinterface, discoverymngr]
+import ../utils/heartbeat
+import ../crypto/crypto
+
+type
+  AutonatService* = ref object of Service
+    registerLoop: Future[void]
+    scheduleInterval: Option[Duration]
+    networkReachability: NetworkReachability
+    t: CountTable[NetworkReachability]
+    autonat: Autonat
+    newStatusHandler: NewStatusHandler
+    rng: ref HmacDrbgContext
+    numPeersToAsk: int
+    maxConfidence: int
+
+
+  NetworkReachability* {.pure.} = enum
+    NotReachable, Reachable, Unknown
+
+  NewStatusHandler* = proc (networkReachability: NetworkReachability): Future[void]  {.gcsafe, raises: [Defect].}
+
+proc new*(
+  T: typedesc[AutonatService],
+  autonat: Autonat,
+  rng: ref HmacDrbgContext,
+  scheduleInterval: Option[Duration] = none(Duration),
+  numPeersToAsk: int = 5,
+  maxConfidence: int = 3): T =
+  return T(
+    scheduleInterval: scheduleInterval,
+    networkReachability: NetworkReachability.Unknown,
+    t: initCountTable[NetworkReachability](),
+    autonat: autonat,
+    rng: rng,
+    numPeersToAsk: numPeersToAsk,
+    maxConfidence: maxConfidence)
+
+proc networkReachability*(self: AutonatService): NetworkReachability {.inline.} =
+  return self.networkReachability
+
+proc handleAnswer(self: AutonatService, ans: NetworkReachability) {.async.} =
+  if ans == NetworkReachability.Unknown:
+    return
+  if ans == self.networkReachability:
+    if self.t[ans] == self.maxConfidence:
+      return
+    self.t.inc(ans)
+  else:
+    if self.t[self.networkReachability] > 0:
+      self.t.inc(self.networkReachability, -1)
+    if self.t[ans] < self.maxConfidence:
+      self.t.inc(ans)
+    if self.t[ans] == self.maxConfidence or self.t[self.networkReachability] == 0:
+      self.networkReachability = ans
+
+  if self.t[self.networkReachability] == self.maxConfidence:
+    if not isNil(self.newStatusHandler):
+      await self.newStatusHandler(self.networkReachability)
+
+  trace "Current status confidence", confidence = $self.t
+  trace "Current status", currentStats = $self.networkReachability
+
+proc askPeer(self: AutonatService, s: Switch, peerId: PeerId): Future[bool] {.async.} =
+  trace "Asking for reachability", peerId = $peerId
+  let (ans, known) =
+    try:
+      let ma = await self.autonat.dialMe(peerId)
+      (NetworkReachability.Reachable, true)
+    except AutonatUnreachableError:
+      (NetworkReachability.NotReachable, true)
+    except AutonatError:
+      (NetworkReachability.Unknown, false)
+  await self.handleAnswer(ans)
+  return known
+
+proc askConnectedPeers(self: AutonatService, switch: Switch) {.async.} =
```

Comment:
As an idea I think we can simplify this by removing `handleAnswer` and not store the states in `t`.

very very roughly
```nim
var states: seq[NetworkState]
for peer in peers:
  if peersToAsk == 0:
    break
  let state = await askPeer(self, switch, peer)
  if state != Unknown:
    peersToAsk -= 1
    states.add(state)

if states.countIt(it == Reachable) > maxConfidence:
   ...
```
and then trigger the `newStatusHandler`.